versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -825,6 +827,119 @@ public final CommitLogEntry fetchFromCommitLog(OP_CONTEXT ctx, Hash hash) {
+    // identifyReferencedAndUnreferencedHeads() knows when it can stop scanning a named-reference's
+    // commit-log. identifyReferencedAndUnreferencedHeads() has to read up to the first commit
+    // _before_ this timestamp to not commit-IDs as "unreferenced".
+    //
+    // Note: keep in mind, that scanAllCommitLogEntries() returns all commits in a
+    // non-deterministic order. Example: if (at least) two commits are added to a branch while this
+    // function is running, the original HEAD of that branch could otherwise be returned as
+    // "unreferenced".
+    long scanStartedAtInMicros = this.commitTimeInMicros();
+
+    try (Stream<CommitLogEntry> scan = scanAllCommitLogEntries()) {
+      scan.forEach(
+          entry -> {
+            Hash parent = entry.getParents().get(0);
+            if (!commits.add(parent)) {
+              // If "parent" has already been added to the set of commits, then it must be a
+              // common ancestor.
+              commonAncestors.add(parent);
+            }
+
+            Hash commitId = entry.getHash();
+            if (!commits.contains(commitId)) {
+              // If the commit-ID is not present in "commits", it must be a HEAD
+              heads.add(commitId);
+            }
+          });
+    }
+
+    // Commits in "commits" that are also contained in "heads" cannot be HEADs.
+    heads.removeIf(commits::contains);
+
+    return HeadsAndCommonAncestors.of(heads, commonAncestors, scanStartedAtInMicros);
+  }
+
+  @Override
+  public ReferencedAndUnreferencedHeads identifyReferencedAndUnreferencedHeads(
+      HeadsAndCommonAncestors headsAndCommonAncestors) throws ReferenceNotFoundException {
+    Map<Hash, Set<NamedRef>> referenced = new HashMap<>();
+    Set<Hash> heads = headsAndCommonAncestors.getHeads();
+    Set<Hash> unreferenced = new HashSet<>(heads);
+
+    // TODO add grace time to account for wall-clock-drift on multiple Nessie instances
+    long stopAtCommitTimeMicros = headsAndCommonAncestors.getScanStartedAtInMicros();
+
+    try (Stream<ReferenceInfo<ByteString>> namedRefs = namedRefs(GetNamedRefsParams.DEFAULT)) {
+      namedRefs.forEach(
+          refInfo -> {
+            boolean canStop = false;
+            try (Stream<CommitLogEntry> logs = commitLog(refInfo.getHash())) {
+              if (!referenced.containsKey(refInfo.getHash())) {
+                // Only need to traverse the commit log from the same commit-ID once.
+
+                boolean identified = false;
+                for (Iterator<CommitLogEntry> logIter = logs.iterator(); logIter.hasNext(); ) {
+                  CommitLogEntry entry = logIter.next();
+
+                  Hash commitId = entry.getHash();
+
+                  if (referenced.containsKey(commitId)) {
+                    // Already saw this commit-ID, can break
+                    break;
+                  }
+
+                  if (heads.contains(commitId)) {
+                    unreferenced.remove(entry.getHash());
```

Comment:
Could we keep track of unreferenced heads by having a log of "remove branch/tag" operations?... that is to avoid scanning?

Reply:
That was what the ref-log was doing.
I thought about another "log" for those operations as well, but it's tricky and would eventually become a "single point of contention" as well. Or using anti-patterns for no-sql databases (think: time-series data)

Reply:
Sorry, I was not clear, I meant keeping a log of "obsolete" HEADs attached to the normal branch HEAD. For each branch we'd have one active HEAD and a chain on previous HEADs (for old deleted branched of the same name).

Reply:
As a side benefit we'd get the previous ref name for unreferenced HEADs, which might be relevant for access control and historical investigation (by end users).